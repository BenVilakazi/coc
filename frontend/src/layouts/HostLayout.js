import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import HamburgerMenu from '../components/Buttons/HamburgerMenu/HamburgerMenu.js';
import Modal from '../components/Modal/Modal.js';
import config from '../config';

const propTypes = {
  left: PropTypes.node.isRequired,
  right: PropTypes.node.isRequired,
  modal: PropTypes.node.isRequired,
};

const {
  smallDesktopWidth,
  largeDesktopWidth,
  extraLargeDesktopWidth,
} = config.breakpoint.hostBreakpoints;

const HostLayoutContainer = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;

  display: flex;
  flex-direction: column;

  .hamburger-container {
    position: absolute;
    right: 17px;
    top: 11px;
    z-index: 10100;
  }

  .host-layout-header {
    display: flex;
    align-items: flex-end;
    height: 50px;
  }

  .host-layout-header h2 {
    position: relative;
    bottom: -1px;
    font-size: 1rem;
    line-height: 0.8rem;
  }

  .components {
    display: flex;
    height: 100%;
  }

  .components .left {
    position: relative;
    display: flex;
    flex-direction: column;
    min-width: 200px;
    background-color: var(--primary-background-color);
    box-shadow: 0px 4px 12px rgba(0, 0, 0, 0.25);
  }

  .components .right {
    display: flex;
    flex-grow: 1;
    position: relative;
    min-width: 440px;
  }

  @media (min-width: ${smallDesktopWidth}) {
    .host-layout-header {
      height: 130px;
    }

    .components .left {
      min-width: 420px;
    }

    .components .host-layout-header h2 {
      font-size: 2.5rem;
      margin-bottom: -4px;
      line-height: 2rem;
    }
  }

  @media (min-width: ${largeDesktopWidth}) {
    .components .left {
      min-width: 700px;
    }

    .components .host-layout-header h2 {
      font-size: 3.5rem;
      margin-bottom: -35px;
      line-height: 7rem;
    }
  }

  @media (min-width: ${extraLargeDesktopWidth}) {
    .host-layout-header {
      height: 200px;
    }

    .components .left {
      min-width: 1100px;
    }

    .components .host-layout-header h2 {
      font-size: 6rem;
      margin-bottom: -25px;
      line-height: 7rem;
    }
  }
`;

const Header = styled.header`
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
`;

function HostLayout({ left, right, modal }) {
  const [hamburgerMenuActive, setHamburgerMenuActive] = useState(false);
  const [showModal, setShowModal] = useState(false);

  function handleModalClick() {
    setHamburgerMenuActive(!hamburgerMenuActive);
    setShowModal(!showModal);
  }

  return (
    <HostLayoutContainer className="primary-background">
      {showModal && <Modal onClickOutside={handleModalClick}>{modal}</Modal>}

      <div className="hamburger-container">
        <HamburgerMenu
          isActive={hamburgerMenuActive}
          onClick={handleModalClick}
        />
      </div>

      <div className="components">
        <div className="left">
          <Header className="host-layout-header">
            <h2>CARDS OF CAROUSAL</h2>
          </Header>
          {left}
        </div>

        <div className="right">{right}</div>
      </div>
    </HostLayoutContainer>
  );
}

HostLayout.propTypes = propTypes;

export default HostLayout;
