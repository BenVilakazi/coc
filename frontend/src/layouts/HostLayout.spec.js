import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import HostLayout from './HostLayout';

describe('HostLayout', () => {
  const leftComponent = <p>Im a left component</p>;
  const rightComponent = <p>Im a right component</p>;
  const modalComponent = <p>Im a modal component</p>;

  // --------------------------------------------------------------------------
  // Rendering Tests
  it('renders the header text', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(screen.getByText('CARDS OF CAROUSAL')).toBeInTheDocument();
  });

  it('renders the text present in the left component', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(screen.getByText('Im a left component')).toBeInTheDocument();
  });

  it('renders the text present in the right component', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(screen.getByText('Im a right component')).toBeInTheDocument();
  });

  it('renders the hamburger menu button', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('does not render the modal until the hamburger menu button is clicked', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(screen.queryByText('Im a modal component')).not.toBeInTheDocument();
  });

  it('renders the modal component when the hamburger menu button is clicked', () => {
    document.body.innerHTML = `
    <div id="modal-root"></div>
  `;

    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    userEvent.click(screen.getByRole('button'));

    expect(screen.getByText('Im a modal component')).toBeTruthy();
  });

  it('does not render the modal when the hamburger menu button is clicked a second time', () => {
    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    userEvent.click(screen.getByRole('button'));
    userEvent.click(screen.getByRole('button'));

    expect(screen.queryByText('Im a modal component')).not.toBeInTheDocument();
  });

  // --------------------------------------------------------------------------
  // PropType Tests

  it('logs an error when attempting to render without the left component', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<HostLayout right={rightComponent} modal={modalComponent} />);

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('logs an error when attempting to render the left component as a non node', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(
      <HostLayout
        left={() => {}}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('logs an error when attempting to render without the right component', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<HostLayout left={leftComponent} modal={modalComponent} />);

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('logs an error when attempting to render the right component as a non node', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(
      <HostLayout
        left={leftComponent}
        right={() => {}}
        modal={modalComponent}
      />,
    );

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('logs an error when attempting to render without the modal component', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<HostLayout left={leftComponent} right={rightComponent} />);

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('logs an error when attempting to render the modal component as a non node', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={() => {}}
      />,
    );

    expect(consoleSpy).toHaveBeenCalled();
  });

  it('does not log any error when all of the needed props are provided', () => {
    const consoleSpy = jest
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(
      <HostLayout
        left={leftComponent}
        right={rightComponent}
        modal={modalComponent}
      />,
    );

    expect(consoleSpy).not.toHaveBeenCalled();
  });

  // --------------------------------------------------------------------------
  // Snapshot Test

  it('matches the snapshot when rendered', () => {
    const tree = renderer
      .create(
        <HostLayout
          left={leftComponent}
          right={rightComponent}
          modal={modalComponent}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
