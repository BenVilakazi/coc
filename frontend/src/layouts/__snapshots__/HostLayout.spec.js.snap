// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostLayout matches the snapshot when rendered 1`] = `
.c1 {
  background-image: url(inactiveHamburger.svg);
  background-size: cover;
  background-color: transparent;
  border: 0;
  width: 25px;
  height: 25px;
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .hamburger-container {
  position: absolute;
  right: 17px;
  top: 11px;
  z-index: 10100;
}

.c0 .host-layout-header {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  height: 50px;
}

.c0 .host-layout-header h2 {
  position: relative;
  bottom: -1px;
  font-size: 1rem;
  line-height: 0.8rem;
}

.c0 .components {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c0 .components .left {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 200px;
  background-color: var(--primary-background-color);
  box-shadow: 0px 4px 12px rgba(0,0,0,0.25);
}

.c0 .components .right {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  position: relative;
  min-width: 440px;
}

.c2 {
  width: 100%;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
}

@media (min-width:680px) {
  .c0 .host-layout-header {
    height: 130px;
  }

  .c0 .components .left {
    min-width: 420px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 2.5rem;
    margin-bottom: -4px;
    line-height: 2rem;
  }
}

@media (min-width:1600px) {
  .c0 .components .left {
    min-width: 700px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 3.5rem;
    margin-bottom: -35px;
    line-height: 7rem;
  }
}

@media (min-width:3200px) {
  .c0 .host-layout-header {
    height: 200px;
  }

  .c0 .components .left {
    min-width: 1100px;
  }

  .c0 .components .host-layout-header h2 {
    font-size: 6rem;
    margin-bottom: -25px;
    line-height: 7rem;
  }
}

<div
  className="c0 primary-background"
>
  <div
    className="hamburger-container"
  >
    <button
      className="c1"
      onClick={[Function]}
    />
  </div>
  <div
    className="components"
  >
    <div
      className="left"
    >
      <header
        className="c2 host-layout-header"
      >
        <h2>
          CARDS OF CAROUSAL
        </h2>
      </header>
      <p>
        Im a left component
      </p>
    </div>
    <div
      className="right"
    >
      <p>
        Im a right component
      </p>
    </div>
  </div>
</div>
`;
